src/lang/checkers/SanityChecker.ml

```diff
@@ -0,0 +1,120 @@
+(*
+  This file is part of scilla.
+
+  Copyright (c) 2018 - present Zilliqa Research Pvt. Ltd.
+  
+  scilla is free software: you can redistribute it and/or modify it under the
+  terms of the GNU General Public License as published by the Free Software
+  Foundation, either version 3 of the License, or (at your option) any later
+  version.
+ 
+  scilla is distributed in the hope that it will be useful, but WITHOUT ANY
+  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ 
+  You should have received a copy of the GNU General Public License along with
+*)
+
+open TypeUtil
+open Syntax
+
+open ContractUtil.MessagePayload
+open MonadUtil
+open Core.Result.Let_syntax
+
+module ScillaSanityChecker
+    (SR : Rep)
+    (ER : sig
+       include Rep
+       val get_type : rep -> PlainTypes.t inferred_type
+     end) = struct
+
+  module SER = SR
+  module EER = ER
+  module EISyntax = ScillaSyntax (SR) (ER)
+  module TU = TypeUtilities (SR) (ER)
+  module SCU = ContractUtil.ScillaContractUtil (SR) (ER)
+
+  open EISyntax
+  open SCU
+
+  (* Basic sanity tests on the contract. *)
+  let contr_sanity (contr : contract) =
+
+    (* Check if there are duplicate entries in "ilist". *)
+    let check_duplicate_ident gloc ilist =
+      let list_count l e =
+        List.fold_left (fun i x -> if get_id x = get_id e then i+1 else i) 0 l
+      in
+      let (e, _) = List.fold_left (fun (e, dlist) i ->
+        let count = list_count ilist i in
+        if count > 1 && not (List.exists (fun x -> get_id x = get_id i) dlist) then
```

Comment:
As far as I can tell the use of `list_count` is to report the error on the last occurrence of the identifier. Can't we just report the error on the second occurrence, and eliminate `list_count`?

Reply:
This is now fixed.